File: тесты/run_all_tests.py
"""
Мастер-скрипт для запуска всех исправленных тестов GENESIS.
Координирует выполнение всех типов тестов и создает сводный отчет.
"""
from __future__ import annotations

import argparse
import json
import os
import subprocess
import sys
import time
from datetime import datetime
from pathlib import Path
from typing import Any, Dict, List, TypedDict


class TestSuite(TypedDict):
    name: str
    file: str
    description: str


class TestSuiteResult(TypedDict):
    name: str
    status: str
    exit_code: int
    output: str
    execution_time: float
    suite_name: str
    description: str
    file_path: str
    error: str


class MasterReport(TypedDict):
    url: str
    timestamp: str
    total_execution_time: float
    test_suites: List[TestSuiteResult]


def run_test_suite(test_file: str, url: str, headless: bool, timeout: int = 30000) -> Dict[str, Any]:
    """Запуск одного набора тестов"""
    print(f"🔄 Запуск {test_file}...")
    
    start_time = time.time()
    
    try:
        # Подготавливаем команду
        cmd = [
            sys.executable,
            test_file,
            "--url", url,
            "--headless", str(int(headless)),
            "--timeout", str(timeout)
        ]
        
        # Запускаем тест
        result = subprocess.run(
            cmd,
            capture_output=True,
            text=True,
            timeout=timeout // 1000 + 60,  # Таймаут в секундах + запас
            cwd=Path(__file__).parent.parent
        )
        
        execution_time = time.time() - start_time
        
        status = "PASS" if result.returncode == 0 else "FAIL"
        
        return {
            "name": Path(test_file).stem,
            "status": status,
            "exit_code": result.returncode,
            "output": result.stdout + result.stderr,
            "execution_time": execution_time,
            "error": "" if status == "PASS" else "Test failed",
            "all_errors": []
        }
    except subprocess.TimeoutExpired:
        execution_time = time.time() - start_time
        return {
            "name": Path(test_file).stem,
            "status": "TIMEOUT",
            "exit_code": -1,
            "output": "Test timed out",
            "execution_time": execution_time,
            "error": "Test timed out",
            "all_errors": ["Test timed out"]
        }
    except Exception as e:
        execution_time = time.time() - start_time
        return {
            "name": Path(test_file).stem,
            "status": "ERROR",
            "exit_code": -1,
            "output": f"Error running test: {str(e)}",
            "execution_time": execution_time,
            "error": str(e),
            "all_errors": [str(e)]
        }


def run_all_tests(url: str, timeout: int = 25000, headless: bool = True) -> Dict[str, Any]:
    """Запуск всех тестов"""
    base_path = Path(__file__).parent
    artifacts_dir = base_path / "_artifacts"
    artifacts_dir.mkdir(exist_ok=True)
    
    ts = datetime.now().strftime("%Y%m%d_%H%M%S")
    start_time = time.time()
    
    # Список всех тестов
    test_suites: List[TestSuite] = [
        TestSuite(
            name="Landing Chrome Test",
            file=str(base_path / "e2e" / "test_landing_chrome.py"),
            description="Базовая проверка терминалов и DOM"
        ),
        TestSuite(
            name="Full Landing Test", 
            file=str(base_path / "e2e" / "test_full_landing.py"),
            description="Полная проверка функциональности лендинга"
        ),
        TestSuite(
            name="Console Monitor",
            file=str(base_path / "e2e" / "test_console_monitor.py"), 
            description="Мониторинг консоли и сетевых запросов"
        ),
        TestSuite(
            name="Functional Tests",
            file=str(base_path / "e2e" / "test_functional.py"),
            description="Функциональные тесты авторизации и кабинета"
        ),
        TestSuite(
            name="JavaScript Unit Tests",
            file=str(base_path / "unit" / "test_js_modules.py"),
            description="Unit тесты JavaScript модулей"
        ),
        TestSuite(
            name="Performance Tests",
            file=str(base_path / "performance" / "test_performance.py"),
            description="Performance, accessibility, SEO тесты"
        )
    ]
    
    results: List[Dict[str, Any]] = []
    total_passed = 0
    total_failed = 0
    total_skipped = 0
    
    print(f"🚀 Запуск полного набора тестов для {url}")
    print(f"📊 Всего тестовых наборов: {len(test_suites)}")
    print("=" * 60)
    
    for i, suite in enumerate(test_suites, 1):
        print(f"[{i}/{len(test_suites)}] {suite['name']}...")
        
        test_file_path = Path(suite["file"])
        
        # Проверяем существование файла
        if not test_file_path.exists():
            result: Dict[str, Any] = {
                "name": Path(suite["file"]).stem,
                "status": "SKIP",
                "exit_code": -1,
                "output": f"Test file not found: {suite['file']}",
                "execution_time": 0,
                "error": "File not found",
                "suite_name": suite["name"],
                "description": suite["description"],
                "file_path": str(suite["file"]),
                "all_errors": [f"File not found: {suite['file']}"]
            }
            total_skipped += 1
        else:
            result = run_test_suite(str(suite["file"]), url, headless, timeout)
            result["suite_name"] = suite["name"]
            result["description"] = suite["description"]
            result["file_path"] = str(suite["file"])
            
            # Подсчет результатов
            if result["status"] == "PASS":
                total_passed += 1
            elif result["status"] != "SKIP":
                total_failed += 1
        
        results.append(result)
        
        # Вывод статуса
        if result["status"] == "PASS":
            print(f"   ✅ PASS")
        elif result["status"] == "SKIP":
            print(f"   ⏭️  SKIP - {result.get('error', 'Unknown')}")
        else:
            print(f"   ❌ FAIL - {result.get('error', 'Unknown')}")
    
    print("=" * 60)
    
    # Сводная статистика
    total_tests = total_passed + total_failed
    success_rate = (total_passed / total_tests * 100) if total_tests > 0 else 0
    total_execution_time = time.time() - start_time
    
    # Сбор всех ошибок
    all_errors: List[str] = []
    for result in results:
        if "all_errors" in result and result["all_errors"]:
            all_errors.extend(result["all_errors"])
        elif result["status"] in ["FAIL", "ERROR", "TIMEOUT"] and "error" in result:
            all_errors.append(f"{result.get('suite_name', result['name'])}: {result['error']}")
    
    # Финальный отчет
    final_report: Dict[str, Any] = {
        "url": url,
        "timestamp": ts,
        "test_type": "full_test_suite",
        "total_execution_time": total_execution_time,
        "summary": {
            "total_suites": len(test_suites),
            "passed": total_passed,
            "failed": total_failed,
            "skipped": total_skipped,
            "success_rate": f"{success_rate:.1f}%"
        },
        "suite_results": results,
        "all_errors": all_errors[:20],  # Первые 20 ошибок
        "artifacts_directory": str(artifacts_dir),
        "status": "PASS" if total_failed == 0 and total_passed > 0 else "FAIL"
    }
    
    # Сохранение сводного отчета
    report_path = artifacts_dir / f"master_report_{ts}.json"
    try:
        report_path.write_text(json.dumps(final_report, ensure_ascii=False, indent=2), encoding="utf-8")
        print(f"📄 Сводный отчет сохранен: {report_path}")
    except Exception as e:
        print(f"⚠️  Ошибка сохранения отчета: {e}")
    
    # Сохранение в REPORT_GENESIS.md
    try:
        genesis_report_path = Path("C:/Users/konfu/Desktop/Sites/Experiment/Experiment1/GENESIS-Website-Clean/REPORT_GENESIS.md")
        if genesis_report_path.exists():
            existing_content = genesis_report_path.read_text(encoding="utf-8")
        else:
            existing_content = "# GENESIS Test Reports\n\n"
        
        new_report = f"""
## Test Run: {ts}
- **URL**: {url}
- **Status**: {'✅ PASS' if final_report['status'] == 'PASS' else '❌ FAIL'}
- **Success Rate**: {final_report['summary']['success_rate']}
- **Passed/Total**: {total_passed}/{len(test_suites)}
- **Execution Time**: {total_execution_time:.2f}s

### Suite Results:
"""
        for result in results:
            status_icon = "✅" if result["status"] == "PASS" else "❌" if result["status"] == "FAIL" else "⏭️"
            new_report += f"- {status_icon} **{result['suite_name']}**: {result['status']} ({result['execution_time']:.2f}s)\n"
        
        if all_errors:
            new_report += "\n### Errors:\n"
            for error in all_errors[:5]:
                new_report += f"- {error}\n"
        
        new_report += "\n---\n"
        
        genesis_report_path.write_text(existing_content + new_report, encoding="utf-8")
        print(f"📝 Отчет добавлен в REPORT_GENESIS.md")
    except Exception as e:
        print(f"⚠️  Ошибка сохранения в REPORT_GENESIS.md: {e}")
    
    return final_report


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Мастер-скрипт для запуска всех тестов GENESIS")
    parser.add_argument("--url", default=os.environ.get("GENESIS_SITE_URL", "https://crypto-processing.net/"))
    parser.add_argument("--timeout", type=int, default=25000)
    parser.add_argument("--headless", type=int, default=1)
    parser.add_argument("--json", action="store_true", help="Вывод только JSON результата")
    args = parser.parse_args()
    
    report = run_all_tests(args.url, args.timeout, bool(args.headless))
    
    if args.json:
        print(json.dumps(report, ensure_ascii=False, indent=2))
    else:
        print(f"""
🎯 ИТОГОВЫЙ РЕЗУЛЬТАТ
URL: {report['url']}
Статус: {'🟢 PASS' if report['status'] == 'PASS' else '🔴 FAIL'}
Успешность: {report['summary']['success_rate']}
Прошло: {report['summary']['passed']}/{report['summary']['total_suites']}
Время выполнения: {report['total_execution_time']:.2f}с

📊 Детали по наборам:""")
        
        for result in report["suite_results"]:
            status_icon = "✅" if result["status"] == "PASS" else "❌" if result["status"] == "FAIL" else "⏭️"
            print(f"  {status_icon} {result['suite_name']} ({result['execution_time']:.2f}с)")
        
        if report["all_errors"]:
            print(f"\n⚠️  Основные ошибки:")
            for error in report["all_errors"][:5]:
                print(f"  • {error}")
    
    sys.exit(0 if report["status"] == "PASS" else 1)

File: тесты/run_all_tests.js
/**
 * Run All Tests
 * Запускает все тесты последовательно и генерирует общий отчет
 */

const path = require('path');
const fs = require('fs').promises;
const { spawn } = require('child_process');

// Импортируем тестовые классы
const CompleteChromeTest = require('./e2e/test_complete_chrome');
const ConsoleMonitor = require('./e2e/test_console_monitor');

class TestRunner {
    constructor() {
        this.results = [];
        this.startTime = new Date();
        this.config = {
            url: process.env.TEST_URL || 'https://crypto-processing.net/',
            headless: process.env.HEADLESS !== 'false',
            timeout: parseInt(process.env.TIMEOUT || '60000')
        };
    }

    async runTest(name, testFunction) {
        console.log('\n' + '='.repeat(60));
        console.log(`🧪 Запуск теста: ${name}`);
        console.log('='.repeat(60));
        
        const startTime = Date.now();
        let status = 'PASSED';
        let error = null;
        let result = null;
        
        try {
            result = await testFunction();
        } catch (e) {
            status = 'FAILED';
            error = e.message;
            console.error(`❌ Тест провален: ${e.message}`);
        }
        
        const duration = Date.now() - startTime;
        
        this.results.push({
            name,
            status,
            duration,
            error,
            result,
            timestamp: new Date().toISOString()
        });
        
        console.log(`⏱️ Время выполнения: ${(duration / 1000).toFixed(2)}с`);
        console.log(`📊 Статус: ${status === 'PASSED' ? '✅ PASSED' : '❌ FAILED'}`);
        
        return { status, duration };
    }

    async runPythonTest(scriptName) {
        return new Promise((resolve, reject) => {
            const pythonPath = process.platform === 'win32' ? 'python' : 'python3';
            const scriptPath = path.join(__dirname, scriptName);
            
            console.log(`🐍 Запуск Python теста: ${scriptName}`);
            
            const child = spawn(pythonPath, [scriptPath, '--url', this.config.url, '--headless', this.config.headless ? '1' : '0'], {
                env: { ...process.env, PYTHONIOENCODING: 'utf-8' }
            });
            
            let output = '';
            let errorOutput = '';
            
            child.stdout.on('data', (data) => {
                const text = data.toString();
                output += text;
                process.stdout.write(text);
            });
            
            child.stderr.on('data', (data) => {
                const text = data.toString();
                errorOutput += text;
                process.stderr.write(text);
            });
            
            child.on('close', (code) => {
                if (code === 0) {
                    resolve({ status: 'PASSED', output });
                } else {
                    resolve({ status: 'FAILED', output, error: errorOutput });
                }
            });
            
            child.on('error', (err) => {
                reject(err);
            });
        });
    }

    async runAllTests() {
        console.log('🚀 GENESIS Complete Test Suite');
        console.log(`📍 URL: ${this.config.url}`);
        console.log(`🖥️ Mode: ${this.config.headless ? 'Headless' : 'Headed'}`);
        console.log(`⏱️ Started: ${this.startTime.toISOString()}`);
        
        // 1. Complete Chrome Test
        await this.runTest('Complete Chrome E2E Test', async () => {
            const test = new CompleteChromeTest();
            return await test.run();
        });
        
        // 2. Console Monitor (30 секунд)
        await this.runTest('Console Monitor (30s)', async () => {
            const monitor = new ConsoleMonitor({
                url: this.config.url,
                duration: 30000,
                headless: this.config.headless
            });
            return await monitor.start();
        });
        
        // 3. Python тесты если доступны
        try {
            await this.runTest('Landing Chrome Test (Python)', async () => {
                return await this.runPythonTest('e2e/test_landing_chrome.py');
            });
        } catch (e) {
            console.log('⚠️ Python тесты пропущены (Python не найден или тесты отсутствуют)');
        }
        
        // 4. Puppeteer тесты модулей
        await this.runTest('Module Tests', async () => {
            return await this.runModuleTests();
        });
        
        // Генерация финального отчета
        await this.generateFinalReport();
    }

    async runModuleTests() {
        const testFiles = [
            'browser/test_console_how_it_works.mjs',
            'browser/test_console_routes.mjs'
        ];
        
        const results = [];
        
        for (const file of testFiles) {
            const filePath = path.join(__dirname, file);
            
            // Проверяем существование файла
            try {
                await fs.access(filePath);
                console.log(`📦 Запуск модульного теста: ${file}`);
                
                // Динамический импорт ESM модуля
                const module = await import(filePath);
                if (module.default && typeof module.default === 'function') {
                    const result = await module.default();
                    results.push({ file, ...result });
                }
            } catch (e) {
                console.log(`⚠️ Тест ${file} пропущен: ${e.message}`);
            }
        }
        
        return results;
    }

    async generateFinalReport() {
        console.log('\n' + '='.repeat(60));
        console.log('📊 ФИНАЛЬНЫЙ ОТЧЕТ');
        console.log('='.repeat(60));
        
        const endTime = new Date();
        const totalDuration = endTime - this.startTime;
        
        const passed = this.results.filter(r => r.status === 'PASSED').length;
        const failed = this.results.filter(r => r.status === 'FAILED').length;
        const total = this.results.length;
        
        const report = {
            url: this.config.url,
            startTime: this.startTime.toISOString(),
            endTime: endTime.toISOString(),
            totalDuration: totalDuration,
            summary: {
                total,
                passed,
                failed,
                successRate: ((passed / total) * 100).toFixed(2) + '%'
            },
            tests: this.results
        };
        
        // Вывод в консоль
        console.log(`\n✅ Пройдено: ${passed}/${total}`);
        console.log(`❌ Провалено: ${failed}/${total}`);
        console.log(`📈 Успешность: ${report.summary.successRate}`);
        console.log(`⏱️ Общее время: ${(totalDuration / 1000).toFixed(2)}с`);
        
        console.log('\n📋 Детали тестов:');
        this.results.forEach(test => {
            const icon = test.status === 'PASSED' ? '✅' : '❌';
            console.log(`${icon} ${test.name} - ${(test.duration / 1000).toFixed(2)}с`);
            if (test.error) {
                console.log(`   └─ Ошибка: ${test.error}`);
            }
        });
        
        // Сохранение отчета
        const timestamp = endTime.toISOString().replace(/[:.]/g, '-');
        const reportDir = path.join(__dirname, 'reports');
        await fs.mkdir(reportDir, { recursive: true });
        
        const reportPath = path.join(reportDir, `final-report-${timestamp}.json`);
        await fs.writeFile(reportPath, JSON.stringify(report, null, 2));
        
        const htmlReport = this.generateHTMLReport(report);
        const htmlPath = path.join(reportDir, `final-report-${timestamp}.html`);
        await fs.writeFile(htmlPath, htmlReport);
        
        console.log('\n📁 Отчеты сохранены:');
        console.log(`   - JSON: ${reportPath}`);
        console.log(`   - HTML: ${htmlPath}`);
        
        // Возвращаем код выхода
        return failed > 0 ? 1 : 0;
    }

    generateHTMLReport(report) {
        const statusColor = report.summary.failed === 0 ? '#4CAF50' : '#f44336';
        
        return `<!DOCTYPE html>
<html lang="ru">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>GENESIS Final Test Report</title>
    <style>
        * { margin: 0; padding: 0; box-sizing: border-box; }
        body { 
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; 
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            min-height: 100vh;
            padding: 20px;
        }
        .container { 
            max-width: 1200px; 
            margin: 0 auto; 
            background: white; 
            border-radius: 20px; 
            box-shadow: 0 20px 60px rgba(0,0,0,0.3);
            overflow: hidden;
        }
        .header {
            background: linear-gradient(135deg, #667eea 0%, #764ba2 100%);
            color: white;
            padding: 40px;
            text-align: center;
        }
        .header h1 {
            font-size: 2.5em;
            margin-bottom: 10px;
        }
        .header .subtitle {
            font-size: 1.2em;
            opacity: 0.9;
        }
        .content {
            padding: 40px;
        }
        .summary-grid {
            display: grid;
            grid-template-columns: repeat(auto-fit, minmax(200px, 1fr));
            gap: 20px;
            margin-bottom: 40px;
        }
        .summary-card {
            background: #f8f9fa;
            padding: 20px;
            border-radius: 10px;
            text-align: center;
            border: 2px solid #e9ecef;
            transition: transform 0.3s, box-shadow 0.3s;
        }
        .summary-card:hover {
            transform: translateY(-5px);
            box-shadow: 0 10px 20px rgba(0,0,0,0.1);
        }
        .summary-card.success {
            border-color: #4CAF50;
            background: #e8f5e9;
        }
        .summary-card.error {
            border-color: #f44336;
            background: #ffebee;
        }
        .summary-card .value {
            font-size: 2.5em;
            font-weight: bold;
            margin-bottom: 10px;
        }
        .summary-card.success .value { color: #4CAF50; }
        .summary-card.error .value { color: #f44336; }
        .summary-card .label {
            color: #666;
            font-size: 0.9em;
            text-transform: uppercase;
            letter-spacing: 1px;
        }
        .tests-section {
            margin-top: 40px;
        }
        .tests-section h2 {
            font-size: 1.8em;
            margin-bottom: 20px;
            color: #333;
        }
        .test-item {
            background: #f8f9fa;
            border-radius: 10px;
            padding: 20px;
            margin-bottom: 15px;
            display: flex;
            align-items: center;
            justify-content: space-between;
            transition: all 0.3s;
        }
        .test-item:hover {
            background: #e9ecef;
        }
        .test-item.passed {
            border-left: 5px solid #4CAF50;
        }
        .test-item.failed {
            border-left: 5px solid #f44336;
        }
        .test-info {
            flex: 1;
        }
        .test-name {
            font-weight: bold;
            font-size: 1.1em;
            margin-bottom: 5px;
        }
        .test-duration {
            color: #666;
            font-size: 0.9em;
        }
        .test-status {
            padding: 8px 16px;
            border-radius: 20px;
            color: white;
            font-weight: bold;
            font-size: 0.9em;
        }
        .test-status.passed {
            background: #4CAF50;
        }
        .test-status.failed {
            background: #f44336;
        }
        .test-error {
            margin-top: 10px;
            padding: 10px;
            background: #ffebee;
            border-radius: 5px;
            color: #c62828;
            font-family: monospace;
            font-size: 0.9em;
        }
        .footer {
            background: #f8f9fa;
            padding: 20px;
            text-align: center;
            color: #666;
        }
        .progress-bar {
            width: 100%;
            height: 30px;
            background: #e9ecef;
            border-radius: 15px;
            overflow: hidden;
            margin: 20px 0;
        }
        .progress-fill {
            height: 100%;
            background: linear-gradient(90deg, #4CAF50, #8BC34A);
            width: ${report.summary.successRate};
            display: flex;
            align-items: center;
            justify-content: center;
            color: white;
            font-weight: bold;
            transition: width 1s ease;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="header">
            <h1>🧪 GENESIS Test Report</h1>
            <div class="subtitle">${report.url}</div>
            <div class="subtitle">${new Date(report.startTime).toLocaleString()}</div>
        </div>
        
        <div class="content">
            <div class="summary-grid">
                <div class="summary-card ${report.summary.failed === 0 ? 'success' : 'error'}">
                    <div class="value">${report.summary.passed}/${report.summary.total}</div>
                    <div class="label">Тестов пройдено</div>
                </div>
                <div class="summary-card">
                    <div class="value" style="color: #2196F3">${report.summary.successRate}</div>
                    <div class="label">Успешность</div>
                </div>
                <div class="summary-card">
                    <div class="value" style="color: #FF9800">${(report.totalDuration / 1000).toFixed(1)}s</div>
                    <div class="label">Время выполнения</div>
                </div>
            </div>
            
            <div class="progress-bar">
                <div class="progress-fill">${report.summary.successRate}</div>
            </div>
            
            <div class="tests-section">
                <h2>Результаты тестов</h2>
                ${report.tests.map(test => `
                    <div class="test-item ${test.status.toLowerCase()}">
                        <div class="test-info">
                            <div class="test-name">${test.name}</div>
                            <div class="test-duration">⏱️ ${(test.duration / 1000).toFixed(2)} секунд</div>
                            ${test.error ? `<div class="test-error">❌ ${test.error}</div>` : ''}
                        </div>
                        <div class="test-status ${test.status.toLowerCase()}">${test.status}</div>
                    </div>
                `).join('')}
            </div>
        </div>
        
        <div class="footer">
            <p>Generated at ${new Date(report.endTime).toLocaleString()}</p>
            <p>GENESIS Test Suite v1.0.0</p>
        </div>
    </div>
</body>
</html>`;
    }
}

// Запуск
if (require.main === module) {
    const runner = new TestRunner();
    runner.runAllTests().then(exitCode => {
        process.exit(exitCode);
    }).catch(error => {
        console.error('💥 Критическая ошибка:', error);
        process.exit(1);
    });
}

module.exports = TestRunner;
